src/ICat.hs

```diff
@@ -1,332 +1,59 @@
-  , ("241", "626262")
-  , ("242", "6c6c6c")
-  , ("243", "767676")
-  , ("244", "808080")
-  , ("245", "8a8a8a")
-  , ("246", "949494")
-  , ("247", "9e9e9e")
-  , ("248", "a8a8a8")
-  , ("249", "b2b2b2")
-  , ("250", "bcbcbc")
-  , ("251", "c6c6c6")
-  , ("252", "d0d0d0")
-  , ("253", "dadada")
-  , ("254", "e4e4e4")
-  , ("255", "eeeeee")
-  ]
-
-
-intToHex :: (Integral a, Show a) => a -> [Char]
-intToHex i =
-  let
-    h = (Numeric.showHex i "")
-  in
-    if i <= 16 then
-     '0' : h
-    else
-      h
-
-
+-- import qualified System.Directory             as Directory
+import qualified System.Environment           as Environment
+
+intToHex :: (Integral a, Show a) => a -> String
+intToHex i
+  | i <= 16   = '0': h
+  | otherwise = h
+  where h = Numeric.showHex i ""
+
+-- you write code like you have to do work upfront
+-- before creating a reference to a computation.
+-- no need for that.
 approximateXtermColor :: (Num a, Ord a) => a -> a
 approximateXtermColor i =
-  let
-    steps = [0x00, 0x5f, 0x87, 0xaf, 0xd7, 0xff]
-    approx' i [] = 0
-    approx' i ((low , high) : steps) =
-      if low <= i && i <= high then
-        let
-          dl = abs (i - low)
-          dh = abs (high - i)
-        in
-          if dl < dh then
-            low
-          else
-            high
-      else
-        approx' i steps
-        
-  in
-    approx' i (zip steps $ tail steps)
-
-
-hex256at :: Picture.Image Picture.PixelRGB8 -> Int -> Int -> [Char]
+  approx' i (zip steps $ tail steps)
+  where steps = [0x00, 0x5f, 0x87, 0xaf, 0xd7, 0xff]
+        approx' _ [] = 0
+        approx' j ((low , high) : steps')
+          | jInRange && dlLower = low
+          | jInRange && not dlLower = high
+          | otherwise = approx' j steps'
+          where dl = abs (j - low)
+                dh = abs (high - j)
+                jInRange = low <= j && j <= high
+                dlLower = dl < dh
+
+hex256at :: Picture.Image Picture.PixelRGB8 -> Int -> Int -> String
 hex256at p x y =
-  let
-    (Picture.PixelRGB8 r g b) = Picture.pixelAt p x y
-    hex = List.intercalate "" $ map (intToHex . approximateXtermColor) [r, g, b]
-  in
-    hex
+  hex
+  where (Picture.PixelRGB8 r g b) = Picture.pixelAt p x y
+        hex = List.intercalate "" $ map (intToHex . approximateXtermColor) [r, g, b]
 
-
-findCodeByHex :: [Char] -> Maybe [Char]
+-- use hlint, it'll tell you how to clean up stuff like I did here.
+findCodeByHex :: String -> Maybe String
 findCodeByHex hex =
-  fmap fst $ List.find (\(c, h) -> h == hex) colorTable
+  fst <$> List.find (\(_, h) -> h == hex) colorTable
 
 
-escapeCodeAt :: Picture.Image Picture.PixelRGB8 -> Int -> Int -> [Char]
+escapeCodeAt :: Picture.Image Picture.PixelRGB8 -> Int -> Int -> String
 escapeCodeAt p x y = escapeCode
   where
     code = Maybe.fromMaybe "0" $ findCodeByHex $ hex256at p x y
     escapeCode = "\ESC[48;5;" ++ code ++ "m"
 
-    
+
+-- this is a bit tower-y and can be split apart and cleaned up
```

Comment:
Yeah, there's really two ideas going on here and the side-effecting part can be split out.